Intrinsic plot sizing

seeM opened the pull request:
This PR adds a new sizing policy, the intrinsic sizing policy, which renders the plot at its "intrinsic" size i.e. the original size of the plot known to the backend.

Addresses https://github.com/posit-dev/positron/issues/4040.

Here's the corresponding Ark PR: https://github.com/posit-dev/ark/pull/470.

### How it works

1. Creating a plot with a backend that supports this feature, will add a new sizing policy. For example, in Python:

    ```python
    import matplotlib.pyplot as plt
    plt.subplots(figsize=(3, 10))[1].plot([0, 1])
    ```

2. The sizing policy's label will contain the source of the sizing information and the original size in the backend's preferred unit of measurement:

    <img width="218" alt="image" src="https://github.com/user-attachments/assets/71e50c9a-0caf-4f5b-bc5b-37db3e246060">

3. You can also save plots at their intrinsic size:
    
    <img width="503" alt="image" src="https://github.com/user-attachments/assets/92c1773c-3533-44ad-98e0-ce14737dac93">

4. If the backend does not support intrinsic sizing, the policy's label will be disabled:

    <img width="432" alt="image" src="https://github.com/user-attachments/assets/17babb0a-8012-4939-8602-e0b8dd756021">

    1. If you switch from a plot with the intrinsic sizing policy selected to a plot that does not support intrinsic sizing, it will fall back to the "auto" sizing policy.
    2. The "Use intrinsic sizing" checkbox will also be disabled in the save modal.

### QA Notes

1. Create Python plots with different sizes and make sure that switching between them updates the plot size and the sizing policy's label, e.g:
    ```python
    import matplotlib.pyplot as plt
    plt.subplots(figsize=(10, 3))[1].plot([0, 1])
    plt.subplots(figsize=(3, 10))[1].plot([0, 1])
    plt.subplots(figsize=(3, 4))[1].plot([0, 1])
    plt.subplots(figsize=(4, 3))[1].plot([0, 1])
    plt.subplots()[1].plot([0, 1])
    ```
2. Switching from a Python plot with the intrinsic sizing policy selected to an R plot should reset to the auto sizing policy.

seeM committed:
allow language runtimes to specify a preferred plot size

seeM committed:
use the plot's size in the placeholder, not the viewport size

seeM committed:
add `get_intrinsic_size` plot comm backend method

seeM committed:
handle get_intrinsic_size in python kernel

seeM committed:
support optional method params in python comms

seeM committed:
update typescript and rust comm generators to support `required` param arg

seeM committed:
fix python render

seeM committed:
wip intrinsic sizing policy

seeM committed:
always provide intrinsic size

seeM committed:
update intrinsic sizing policy label

seeM committed:
wip - debugging

seeM committed:
extract render function

seeM committed:
restore figures to intrinsic size

seeM committed:
wip placeholder for intrinsic sizing policy

seeM committed:
support optional results

seeM committed:
fix optional result comm generator; make intrinsic size result optional

seeM committed:
Revert "support optional results"

This reverts commit bd76809fa7da1d4c83f59bb7c5b953a61b7ccccd.

seeM committed:
fix plot client given optional result

seeM committed:
make data explorer params required, since generated comms were not supporting optional params anyway

seeM committed:
fix rust generated comm method params

seeM committed:
fix python plots comm handler

seeM committed:
make `intrinsic_size` required

seeM committed:
cleanup optional content descriptor comm generation

seeM committed:
revert unneeded changes

seeM committed:
revert `preferred_size`

seeM committed:
refactor logic to name getter

seeM committed:
remove viewport size from placeholder text

seeM committed:
when a plot is selected, update the intrinsic sizing policy

seeM committed:
cleanup sizing policy component

seeM committed:
move state from intrinsic sizing policy to plot client instance

seeM committed:
reworking

seeM committed:
fix: disable intrinsic policy when intrinsic size is unknown

seeM committed:
extract function

seeM committed:
cleanup

seeM committed:
fix bug where previews were rerendering the main plot

This was previously masked since we didn't declare
useEffect dependencies, so the setUri onDidCompleteRender
would trigger the effect which would re-render with the
policy's sizing.

seeM committed:
add intrinsic sizing checkbox to preview

seeM committed:
comments

seeM committed:
finalize save plot modal

seeM committed:
cleanup plot comm backend spec

seeM committed:
python intrinsic size plot unit tests

seeM commented:
Smoke test is failing since https://github.com/posit-dev/ark/pull/470 is required.

nstrayer commented on src/vs/workbench/services/languageRuntime/common/languageRuntimePlotClient.ts:
```diff
@@ -485,18 +534,18 @@ export class PlotClientInstance extends Disposable implements IPositronPlotClien
 
 		// If we have never rendered this plot, we can't process any updates
 		// yet.
-		if (!this._currentRender && !this._lastRender) {
+		const render = this._currentRender?.renderRequest ?? this._lastRender;
+		if (!render) {
 			return Promise.reject(new Error('Cannot update plot before it has been rendered'));
 		}
 
 		// Use the dimensions of the last or current render request to determine
 		// the size and DPI of the plot to update.
-		const height = this._currentRender?.renderRequest.height ??
-			this._lastRender?.height;
-		const width = this._currentRender?.renderRequest.width ??
-			this._lastRender?.width;
-		const pixel_ratio = this._currentRender?.renderRequest.pixel_ratio ??
-			this._lastRender?.pixel_ratio;
+		const sizeInt = render.size && {
+			height: Math.floor(render.size.height),
+			width: Math.floor(render.size.width)
+		};
+		const pixel_ratio = render.pixel_ratio;
```
This is much nicer!

nstrayer reviewed (dismissed):
This is awesome (and herculean). 

I don't have enough experience with the coms side of things to speak to best practices but everything looks good to me. I also added some notes about typescript/react patterns. Feel free to pick and choose what you want to take from it, nothing was a deal breaker.

seeM commented on positron/comms/plot-backend-openrpc.json:
```diff
@@ -62,11 +88,39 @@
+				"description": "The size of a plot",
+				"properties": {
+					"height": {
+						"description": "The plot's height, in pixels",
```
This one is still in pixels, and `intrinsic_size` is in `plot_unit`.

seeM commented on positron/comms/data_explorer-backend-openrpc.json:
```diff
@@ -47,7 +47,7 @@
 					"name": "start_index",
 					"description": "Index (starting from zero) of first result to fetch (for paging)",
-					"required": false,
+					"required": true,
```
Yuup. These two changes are because `generate-comms.ts` wasn't previously handling this field so all of our code assumes that these are actually required.

jmcphers replied:
I also wish JSON could have comments. :-(

seeM commented on src/vs/workbench/contrib/positronPlots/browser/components/dynamicPlotInstance.tsx:
```diff
@@ -39,49 +43,37 @@ export const DynamicPlotInstance = (props: DynamicPlotInstanceProps) => {
 
 	useEffect(() => {
-		const ratio = window.devicePixelRatio;
+		const ratio = DOM.getActiveWindow().devicePixelRatio;
```
Good point! I'm happy to merge this after and fix conflicts here.

seeM committed:
extract `activeWindow`

seeM committed:
use `disposableTimeout`

seeM committed:
rename

seeM commented on src/vs/workbench/contrib/positronPlots/browser/modalDialogs/savePlotModalDialog.tsx:
```diff
@@ -233,6 +249,26 @@ const SavePlotModalDialog = (props: SavePlotModalDialogProps) => {
+			displayWidth *= dpi.value;
+			displayHeight *= dpi.value;
+		} else if (props.plotIntrinsicSize.unit !== PlotUnit.Pixels) {
+			props.logService.error('Unknown unit');
```
Ah good point. I've just removed the logs for now.

seeM committed:
remove logs on render

seeM committed:
use ternary

seeM commented on src/vs/workbench/contrib/positronPlots/browser/modalDialogs/savePlotModalDialog.tsx:
```diff
@@ -349,6 +387,14 @@ const SavePlotModalDialog = (props: SavePlotModalDialogProps) => {
 							</div>
 						</div>
+						<div className='use-intrinsic-size'>
+							{props.plotIntrinsicSize && <Checkbox
+								label={(() => localize(
+									'positron.savePlotModalDialog.useIntrinsicSize',
+									"Use intrinsic size"
+								))()}
+								onChanged={checked => setUseIntrinsicSize(checked)} />}
```
Good point, will do!

seeM commented:
Thanks for the review @nstrayer! I've addressed all of your comments. Just need to merge the Ark PR, bump Ark, rerun the smoke tests, before we can merge.

seeM committed:
set initial `enableIntrinsicSize` using selected policy; default width and height to 100 (behavior before this PR)

seeM committed:
fix bad merge conflict resolution

jmcphers commented on src/vs/workbench/contrib/positronPlots/browser/modalDialogs/savePlotModalDialog.tsx:
```diff
@@ -197,6 +206,9 @@ const SavePlotModalDialog = (props: SavePlotModalDialogProps) => {
 					props.savePlotCallback({ uri: plotResult.uri, path: filePath });
 				})
+				.catch((error) => {
+					props.logService.error('Error saving plot:', error);
```
Should this error be displayed to the user?

seeM replied:
Yes!

jmcphers commented on src/vs/workbench/services/languageRuntime/common/languageRuntimePlotClient.ts:
```diff
@@ -259,26 +276,44 @@ export class PlotClientInstance extends Disposable implements IPositronPlotClien
+	 *
+	 * @returns A promise that resolves to the intrinsic size of the plot, if known.
+	 */
+	public async getIntrinsicSize(): Promise<IntrinsicSize | undefined> {
```
It's possible for this to be called quickly in succession since it's invoked during render. To avoid the possibility of stacking intrinsic size requests, we should probably check for an in-flight request here. Maybe use a `DeferredPromise` we can return?

seeM replied:
Good point, will do. Is there any reason to prefer a `DeferredPromise` over a `Promise` e.g. like:

```typescript
	public getIntrinsicSize(): Promise<IntrinsicSize | undefined> {
		// If there's already an in-flight request, return its response.
		if (this._currentIntrinsicSize) {
			return this._currentIntrinsicSize;
		}
		this._currentIntrinsicSize = this._comm.getIntrinsicSize()
			.then((intrinsicSize) => {
				this._intrinsicSize = intrinsicSize;
				this._receivedIntrinsicSize = true;
				this._didSetIntrinsicSizeEmitter.fire(intrinsicSize);
				return intrinsicSize;
			})
			.finally(() => {
				this._currentIntrinsicSize = undefined;
			});
		return this._currentIntrinsicSize;
	}
```

I've tended to avoid `DeferredPromise` where possible since I tend to forget to try-catch and reject properly.

jmcphers commented on src/vs/workbench/services/positronPlots/common/sizingPolicyIntrinsic.ts:
```diff
@@ -0,0 +1,45 @@
+				break;
+		}
+
+		return `${intrinsicSize.source} (${intrinsicSize.width}${unit}×${intrinsicSize.height}${unit})`;
```
This should be a localized string with placeholders (units should also be localized)

seeM committed:
address review comments

seeM committed:
bump ark to v0.1.126
